Closing Incognito Tab closes New Tabs under Normal Tabs and redirects to 1st tab

### Bromite version

Version: `93.0.4577.83`
Arch: `arm64`
Android version: 10
Device model: Nokia 6.1 Plus (DRG)

### Flags
These are the flags changed under `chrome://flags`:
- Enable IPv6 Probing (_Enabled_)
- Parallel Downloading (_Enabled_)
- Download auto-resumption native (_Enabled_)
- tab engagement metrics (_Disabled_)
- Support for HTTPSSVC records in DNS (_Disabled_)

### Is this bug about the SystemWebView?
No

### Is the bug reproducible with latest version?
Yes

### Can the bug be reproduced with corresponding Chromium version?
No

### Allow ads for the site; is the bug caused by the adblocker?
No

### Is the bug a crash?
No

### Describe the bug
- I've some tabs opened for reference. Along with that I usually keep a newtab in the forefront of other previous tabs.  
- For random things, I use incognito tabs.  
- When I close them, once my research is done, it brings me to the 1st normal tab closing the newtab which'd be in the forefront.
- I literally feel it annoying as I'm supposed to open the newtab again every time I close an incognito tab, which doesn't happen in chromium.

### Steps to reproduce the bug

Steps to reproduce the bug:
1. Have some tabs open under normal tabs
2. Have **NewTab** opened at last.
3. Open an **incognito tab** and do your research. 
4. Close the incognito tab my any means. Either close from the notification bar/clear the RAM, or systematically close the incognito tab by clicking tab and **x** button(this doesn't close the newtab but highlights the first tab).

### Expected behavior
- It should not touch the newtab i.e., opened. 
- When I close the incognito tab, you should be redirected to the last tab atleast, not the beginning of the tab(1st tab).


### Recording
**Incognito recording has been turned black by default.**   
Here's the recording.(Initially, it showed no bug like as I said before, but on doing the same for the 2nd time, it happened.)  

**LINK:**   
https://user-images.githubusercontent.com/69745008/112744179-5bfe9780-8fbb-11eb-9519-5e94c6988081.mp4  
(Above link didn't seem to work, so, compressed the video into a zip uploaded below.)  
[Bug in Bromite.zip](https://github.com/bromite/bromite/files/6216933/Bug.in.Bromite.zip)

do you have the new homepage option enabled with "custom web address" to nothing?

with those settings I detect this crash, but maybe it has nothing to do with it.
```
04-10 07:58:06.886  9412  9412 W System.err: java.lang.AssertionError: Homepage should be disabled for a null URL
04-10 07:58:06.886  9412  9412 W System.err:    at org.chromium.chrome.browser.metrics.LaunchMetrics.recordHomePageLaunchMetrics(LaunchMetrics.java:106)
04-10 07:58:06.886  9412  9412 W System.err:    at org.chromium.chrome.browser.init.ProcessInitializationHandler$2$1.run(ProcessInitializationHandler.java:303)
04-10 07:58:06.886  9412  9412 W System.err:    at android.os.MessageQueue.nativePollOnce(Native Method)
04-10 07:58:06.886  9412  9412 W System.err:    at android.os.MessageQueue.next(MessageQueue.java:326)
04-10 07:58:06.886  9412  9412 W System.err:    at android.os.Looper.loop(Looper.java:160)
04-10 07:58:06.886  9412  9412 W System.err:    at android.app.ActivityThread.main(ActivityThread.java:6669)
04-10 07:58:06.886  9412  9412 W System.err:    at java.lang.reflect.Method.invoke(Native Method)
04-10 07:58:06.886  9412  9412 W System.err:    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
04-10 07:58:06.886  9412  9412 W System.err:    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
04-10 07:58:06.887  9412  9468 I cr_ProcessInitHandler: Attempting to upload 1 accumulated crash dumps.
04-10 07:58:06.887  9412  9468 I cr_MinidumpJobService: Scheduling upload of all pending minidumps.
04-10 07:58:06.890  9412  9462 I cr_LogcatExtraction: Trying to extract logcat for minidump 6fc23d49-8f39-41fe-8307-be31e20aaa42.dmp9242.
04-10 07:58:06.898  9412  9468 E cr_PackageMetrics: Could not parse UUID 1001-1409
04-10 07:58:06.898  9412  9468 E cr_PackageMetrics: java.lang.IllegalArgumentException: Invalid UUID string: 1001-1409
04-10 07:58:06.898  9412  9468 E cr_PackageMetrics:     at java.util.UUID.fromString(UUID.java:194)
04-10 07:58:06.898  9412  9468 E cr_PackageMetrics:     at org.chromium.chrome.browser.metrics.PackageMetrics.getPackageStatsForAndroidO(PackageMetrics.java:55)
04-10 07:58:06.898  9412  9468 E cr_PackageMetrics:     at org.chromium.chrome.browser.metrics.PackageMetrics.recordPackageStats(PackageMetrics.java:84)
04-10 07:58:06.898  9412  9468 E cr_PackageMetrics:     at org.chromium.chrome.browser.init.ProcessInitializationHandler$11.doInBackground(ProcessInitializationHandler.java:473)
04-10 07:58:06.898  9412  9468 E cr_PackageMetrics:     at org.chromium.chrome.browser.init.ProcessInitializationHandler$11.doInBackground(ProcessInitializationHandler.java:430)
04-10 07:58:06.898  9412  9468 E cr_PackageMetrics:     at org.chromium.base.task.AsyncTask$1.call(AsyncTask.java:106)
04-10 07:58:06.898  9412  9468 E cr_PackageMetrics:     at java.util.concurrent.FutureTask.run(FutureTask.java:266)
04-10 07:58:06.898  9412  9468 E cr_PackageMetrics:     at org.chromium.base.task.AsyncTask$NamedFutureTask.run(AsyncTask.java:404)
04-10 07:58:06.910  9412  9412 F chromium: [FATAL:jni_android.cc(306)] Please include Java exception stack in crash report
```

@csagan5 in the situation described by the user 
under memory pressure  with

`adb shell am send-trim-memory org.chromium.chrome.stable COMPLETE`

I had another crash:

```
04-10 09:00:35.338 13146 13146 E AndroidRuntime: java.lang.AssertionError: Destroy called before removed from window
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at org.chromium.chrome.browser.ntp.IncognitoNewTabPage.destroy(IncognitoNewTabPage.java:151)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at org.chromium.chrome.browser.ui.native_page.FrozenNativePage.freeze(FrozenNativePage.java:32)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at org.chromium.chrome.browser.tab.TabImpl.freezeNativePage(TabImpl.java:429)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at org.chromium.chrome.browser.native_page.NativePageAssassin.freeze(NativePageAssassin.java:87)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at org.chromium.chrome.browser.native_page.NativePageAssassin.freezeAllHiddenPages(NativePageAssassin.java:81)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at org.chromium.chrome.browser.ChromeTabbedActivity.onTrimMemory(ChromeTabbedActivity.java:2322)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at android.app.ActivityThread.handleTrimMemory(ActivityThread.java:5431)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at android.app.ActivityThread.access$900(ActivityThread.java:199)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at android.app.ActivityThread$ApplicationThread.lambda$scheduleTrimMemory$0(ActivityThread.java:1419)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at android.app.-$$Lambda$ActivityThread$ApplicationThread$tUGFX7CUhzB4Pg5wFd5yeqOnu38.accept(Unknown Source:8)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at com.android.internal.util.function.pooled.PooledLambdaImpl.doInvoke(PooledLambdaImpl.java:258)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at com.android.internal.util.function.pooled.PooledLambdaImpl.invoke(PooledLambdaImpl.java:182)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at com.android.internal.util.function.pooled.OmniFunction.run(OmniFunction.java:77)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:949)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at android.view.Choreographer.doCallbacks(Choreographer.java:761)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at android.view.Choreographer.doFrame(Choreographer.java:698)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:935)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:873)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:99)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:193)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:6669)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
04-10 09:00:35.338 13146 13146 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
```

maybe it has nothing to do with it and maybe chromium also behaves like this

--- a/build/patches/Keep-empty-tabs-between-sessions.patch
+++ b/build/patches/Keep-empty-tabs-between-sessions.patch
@@ -1,0 +1,45 @@
+From: uazo <uazo@users.noreply.github.com>
+Date: Wed, 27 Oct 2021 10:57:09 +0000
+Subject: Keep empty tabs between sessions
+
+---
+ .../chrome/browser/tabmodel/TabPersistentStore.java      | 9 ---------
+ .../browser/tab/state/CriticalPersistedTabData.java      | 4 ----
+ 2 files changed, 13 deletions(-)
+
+diff --git a/chrome/android/java/src/org/chromium/chrome/browser/tabmodel/TabPersistentStore.java b/chrome/android/java/src/org/chromium/chrome/browser/tabmodel/TabPersistentStore.java
+--- a/chrome/android/java/src/org/chromium/chrome/browser/tabmodel/TabPersistentStore.java
++++ b/chrome/android/java/src/org/chromium/chrome/browser/tabmodel/TabPersistentStore.java
+@@ -734,15 +734,6 @@ public class TabPersistentStore {
+                     .createFrozenTab(tabState, serializedCriticalPersistedTabData, tabToRestore.id,
+                             isIncognito, restoredIndex);
+         } else {
+-            if (UrlUtilities.isNTPUrl(tabToRestore.url) && !setAsActive
+-                    && !tabToRestore.fromMerge) {
+-                Log.i(TAG, "Skipping restore of non-selected NTP.");
+-                return;
+-            } else if (TextUtils.isEmpty(tabToRestore.url)) {
+-                Log.i(TAG, "Skipping restore of empty Tabs.");
+-                return;
+-            }
+-
+             Log.w(TAG, "Failed to restore TabState; creating Tab with last known URL.");
+             Tab fallbackTab = mTabCreatorManager.getTabCreator(isIncognito)
+                                       .createNewTab(new LoadUrlParams(tabToRestore.url),
+diff --git a/chrome/browser/tab/java/src/org/chromium/chrome/browser/tab/state/CriticalPersistedTabData.java b/chrome/browser/tab/java/src/org/chromium/chrome/browser/tab/state/CriticalPersistedTabData.java
+--- a/chrome/browser/tab/java/src/org/chromium/chrome/browser/tab/state/CriticalPersistedTabData.java
++++ b/chrome/browser/tab/java/src/org/chromium/chrome/browser/tab/state/CriticalPersistedTabData.java
+@@ -438,10 +438,6 @@ public class CriticalPersistedTabData extends PersistedTabData {
+         if (getUrl() == null || TextUtils.isEmpty(getUrl().getSpec())) {
+             return false;
+         }
+-        if (UrlUtilities.isNTPUrl(getUrl().getSpec()) && !mTab.canGoBack()
+-                && !mTab.canGoForward()) {
+-            return false;
+-        }
+         if (isTabUrlContentScheme(getUrl().getSpec())) {
+             return false;
+         }
+-- 
+2.17.1
+

